File: scripts/generate_26bp_whitelists.py
#!/usr/bin/env python3
import argparse
import csv
import re
import sys
import xml.etree.ElementTree as ET
from itertools import product
from pathlib import Path
from zipfile import ZipFile


VALID_BC_RE = re.compile(r"^[ACGT]{10}$")


def reverse_complement(seq):
    return seq.translate(str.maketrans("ACGT", "TGCA"))[::-1]


def read_plain_barcodes(path):
    barcodes = []
    with open(path, "r", encoding="utf-8-sig") as handle:
        sample = handle.read(4096)
        handle.seek(0)
        if "," in sample or "\t" in sample:
            dialect = csv.Sniffer().sniff(sample, delimiters=",\t")
            reader = csv.DictReader(handle, dialect=dialect)
            if reader.fieldnames:
                seq_col = next(
                    (
                        col
                        for col in reader.fieldnames
                        if col and col.strip().lower() in {"sequence", "seq", "barcode", "cellbarcode", "序列"}
                    ),
                    None,
                )
                if seq_col:
                    for row in reader:
                        value = (row.get(seq_col) or "").strip().upper()
                        if value:
                            barcodes.append(value)
                    return barcodes
            handle.seek(0)
        for line in handle:
            value = line.strip().split()[0].upper() if line.strip() else ""
            if value and value not in {"SEQUENCE", "SEQ", "BARCODE", "序列"}:
                barcodes.append(value)
    return barcodes


def _xlsx_shared_strings(zf):
    if "xl/sharedStrings.xml" not in zf.namelist():
        return []
    root = ET.fromstring(zf.read("xl/sharedStrings.xml"))
    ns = {"x": "http://schemas.openxmlformats.org/spreadsheetml/2006/main"}
    values = []
    for item in root.findall("x:si", ns):
        texts = [node.text or "" for node in item.findall(".//x:t", ns)]
        values.append("".join(texts))
    return values


def read_xlsx_barcodes(path, sequence_column_name="序列"):
    with ZipFile(path) as zf:
        shared_strings = _xlsx_shared_strings(zf)
        sheet_name = "xl/worksheets/sheet1.xml"
        root = ET.fromstring(zf.read(sheet_name))
    ns = {"x": "http://schemas.openxmlformats.org/spreadsheetml/2006/main"}

    rows = []
    for row in root.findall(".//x:sheetData/x:row", ns):
        values = {}
        for cell in row.findall("x:c", ns):
            ref = cell.attrib.get("r", "")
            col = re.sub(r"\d+", "", ref)
            value_node = cell.find("x:v", ns)
            if value_node is None:
                value = ""
            elif cell.attrib.get("t") == "s":
                value = shared_strings[int(value_node.text)]
            else:
                value = value_node.text or ""
            values[col] = str(value).strip()
        rows.append(values)

    if not rows:
        return []
    header = rows[0]
    seq_col = next((col for col, name in header.items() if name.strip() == sequence_column_name), None)
    if seq_col is None:
        seq_col = "B"

    return [row.get(seq_col, "").strip().upper() for row in rows[1:] if row.get(seq_col, "").strip()]


def read_10bp_barcodes(path):
    path = Path(path)
    if path.suffix.lower() == ".xlsx":
        barcodes = read_xlsx_barcodes(path)
    else:
        barcodes = read_plain_barcodes(path)

    bad_len = [bc for bc in barcodes if len(bc) != 10]
    bad_base = [bc for bc in barcodes if len(bc) == 10 and not VALID_BC_RE.match(bc)]
    if bad_len:
        raise ValueError(f"Found non-10bp barcode examples: {bad_len[:10]}")
    if bad_base:
        raise ValueError(f"Found barcode with non-ACGT bases examples: {bad_base[:10]}")
    if len(barcodes) != len(set(barcodes)):
        duplicates = [bc for bc, n in csv_counter(barcodes).items() if n > 1]
        raise ValueError(f"Found duplicate barcodes examples: {duplicates[:10]}")
    return barcodes


def csv_counter(values):
    counts = {}
    for value in values:
        counts[value] = counts.get(value, 0) + 1
    return counts


def write_whitelists(barcodes, middle_3p, middle_5p, out_3p, out_5p):
    barcodes_5p = [reverse_complement(bc) for bc in barcodes]
    if len(barcodes_5p) != len(set(barcodes_5p)):
        raise ValueError("5p barcode reverse-complement list has duplicates.")

    count_3p = 0
    with open(out_3p, "w", encoding="utf-8") as handle:
        for left, right in product(barcodes, repeat=2):
            handle.write(f"{left}{middle_3p}{right}\n")
            count_3p += 1

    count_5p = 0
    with open(out_5p, "w", encoding="utf-8") as handle:
        for left, right in product(barcodes_5p, repeat=2):
            handle.write(f"{left}{middle_5p}{right}\n")
            count_5p += 1

    return count_3p, count_5p


def main():
    parser = argparse.ArgumentParser(description="Expand a 10bp barcode list to Strint2 26bp 3p/5p whitelists.")
    parser.add_argument("--barcode-list-10bp", required=True)
    parser.add_argument("--out-3p", required=True)
    parser.add_argument("--out-5p", required=True)
    parser.add_argument("--middle-3p", default="GGTAGC")
    parser.add_argument("--middle-5p", default="GGAAGG")
    args = parser.parse_args()

    barcodes = read_10bp_barcodes(args.barcode_list_10bp)
    Path(args.out_3p).parent.mkdir(parents=True, exist_ok=True)
    Path(args.out_5p).parent.mkdir(parents=True, exist_ok=True)
    count_3p, count_5p = write_whitelists(
        barcodes=barcodes,
        middle_3p=args.middle_3p,
        middle_5p=args.middle_5p,
        out_3p=args.out_3p,
        out_5p=args.out_5p,
    )

    print(f"Input 10bp barcode count: {len(barcodes)}")
    print(f"3p whitelist: {args.out_3p} ({count_3p} records)")
    print(f"5p whitelist: {args.out_5p} ({count_5p} records)")
    print(f"First 3 raw barcodes: {barcodes[:3]}")
    print(f"First 3 5p RC barcodes: {[reverse_complement(bc) for bc in barcodes[:3]]}")


if __name__ == "__main__":
    try:
        main()
    except Exception as exc:
        print(f"[ERROR] {exc}", file=sys.stderr)
        sys.exit(1)

File: src/bin/generate_26bp_whitelists.rs
use std::collections::{HashMap, HashSet};
use std::fs::{self, File};
use std::io::{Read, Write};
use std::path::{Path, PathBuf};

use anyhow::{bail, Context, Result};
use clap::Parser;
use flora::barcode::reverse_complement;
use zip::ZipArchive;

#[derive(Debug, Parser)]
#[command(version, about = "Expand 10bp barcodes to Flora 26bp 3p/5p whitelists")]
struct Cli {
    #[arg(long = "barcode-list-10bp")]
    barcode_list_10bp: PathBuf,

    #[arg(long = "out-3p")]
    out_3p: PathBuf,

    #[arg(long = "out-5p")]
    out_5p: PathBuf,

    #[arg(long = "middle-3p", default_value = "GGTAGC")]
    middle_3p: String,

    #[arg(long = "middle-5p", default_value = "GGAAGG")]
    middle_5p: String,
}

pub fn main() -> Result<()> {
    let cli = Cli::parse();
    let barcodes = read_10bp_barcodes(&cli.barcode_list_10bp)?;
    if let Some(parent) = cli.out_3p.parent() {
        fs::create_dir_all(parent)?;
    }
    if let Some(parent) = cli.out_5p.parent() {
        fs::create_dir_all(parent)?;
    }
    let (count_3p, count_5p) = write_whitelists(
        &barcodes,
        &cli.middle_3p,
        &cli.middle_5p,
        &cli.out_3p,
        &cli.out_5p,
    )?;
    println!("Input 10bp barcode count: {}", barcodes.len());
    println!(
        "3p whitelist: {} ({count_3p} records)",
        cli.out_3p.display()
    );
    println!(
        "5p whitelist: {} ({count_5p} records)",
        cli.out_5p.display()
    );
    println!(
        "First 3 raw barcodes: {:?}",
        &barcodes[..barcodes.len().min(3)]
    );
    let first_rc: Vec<_> = barcodes
        .iter()
        .take(3)
        .map(|bc| reverse_complement(bc))
        .collect();
    println!("First 3 5p RC barcodes: {first_rc:?}");
    Ok(())
}

fn read_10bp_barcodes(path: &Path) -> Result<Vec<String>> {
    let barcodes = if path
        .extension()
        .and_then(|x| x.to_str())
        .is_some_and(|x| x.eq_ignore_ascii_case("xlsx"))
    {
        read_xlsx_barcodes(path)?
    } else {
        read_plain_barcodes(path)?
    };
    validate_barcodes(barcodes)
}

fn read_plain_barcodes(path: &Path) -> Result<Vec<String>> {
    let text = fs::read_to_string(path).with_context(|| format!("read {}", path.display()))?;
    let first_line = text
        .lines()
        .find(|line| !line.trim().is_empty())
        .unwrap_or("");
    let delimiter = if first_line.contains(',') {
        Some(',')
    } else if first_line.contains('\t') {
        Some('\t')
    } else {
        None
    };
    if let Some(delimiter) = delimiter {
        let mut lines = text.lines();
        let header_line = lines.next().unwrap_or("");
        let headers: Vec<String> = header_line
            .split(delimiter)
            .map(|x| x.trim().trim_matches('\u{feff}').to_ascii_lowercase())
            .collect();
        let seq_idx = headers.iter().position(|h| {
            matches!(
                h.as_str(),
                "sequence" | "seq" | "barcode" | "cellbarcode" | "序列"
            )
        });
        if let Some(idx) = seq_idx {
            return Ok(lines
                .filter_map(|line| line.split(delimiter).nth(idx))
                .map(|x| x.trim().to_ascii_uppercase())
                .filter(|x| !x.is_empty())
                .collect());
        }
    }
    Ok(text
        .lines()
        .filter_map(|line| line.split_whitespace().next())
        .map(|x| x.trim().trim_matches('\u{feff}').to_ascii_uppercase())
        .filter(|x| !x.is_empty() && !matches!(x.as_str(), "SEQUENCE" | "SEQ" | "BARCODE" | "序列"))
        .collect())
}

fn read_xlsx_barcodes(path: &Path) -> Result<Vec<String>> {
    let file = File::open(path)?;
    let mut zip = ZipArchive::new(file)?;
    let shared_strings = read_shared_strings(&mut zip)?;
    let mut sheet = String::new();
    zip.by_name("xl/worksheets/sheet1.xml")?
        .read_to_string(&mut sheet)?;
    let rows = parse_sheet_rows(&sheet, &shared_strings);
    if rows.is_empty() {
        return Ok(Vec::new());
    }
    let header = &rows[0];
    let seq_col = header
        .iter()
        .find_map(|(col, value)| (value.trim() == "序列").then_some(col.clone()))
        .unwrap_or_else(|| "B".to_string());
    Ok(rows
        .iter()
        .skip(1)
        .filter_map(|row| row.get(&seq_col))
        .map(|x| x.trim().to_ascii_uppercase())
        .filter(|x| !x.is_empty())
        .collect())
}

fn read_shared_strings(zip: &mut ZipArchive<File>) -> Result<Vec<String>> {
    let Ok(mut member) = zip.by_name("xl/sharedStrings.xml") else {
        return Ok(Vec::new());
    };
    let mut xml = String::new();
    member.read_to_string(&mut xml)?;
    Ok(xml
        .split("<si>")
        .skip(1)
        .map(|item| {
            item.split("<t")
                .skip(1)
                .filter_map(|part| part.split_once('>')?.1.split_once("</t>").map(|x| x.0))
                .collect::<String>()
        })
        .collect())
}

fn parse_sheet_rows(xml: &str, shared_strings: &[String]) -> Vec<HashMap<String, String>> {
    let mut rows = Vec::new();
    for row_xml in xml.split("<row").skip(1) {
        let Some(row_body) = row_xml.split_once("</row>").map(|x| x.0) else {
            continue;
        };
        let mut row = HashMap::new();
        for cell_xml in row_body.split("<c ").skip(1) {
            let cell_head = cell_xml.split_once('>').map(|x| x.0).unwrap_or("");
            let col = attr_value(cell_head, "r")
                .unwrap_or_default()
                .chars()
                .filter(|c| c.is_ascii_alphabetic())
                .collect::<String>();
            if col.is_empty() {
                continue;
            }
            let raw = cell_xml
                .split_once("<v>")
                .and_then(|x| x.1.split_once("</v>"))
                .map(|x| x.0)
                .unwrap_or("");
            let value = if cell_head.contains("t=\"s\"") {
                raw.parse::<usize>()
                    .ok()
                    .and_then(|idx| shared_strings.get(idx))
                    .cloned()
                    .unwrap_or_default()
            } else {
                raw.to_string()
            };
            row.insert(col, value);
        }
        rows.push(row);
    }
    rows
}

fn attr_value(text: &str, name: &str) -> Option<String> {
    let needle = format!("{name}=\"");
    let rest = text.split_once(&needle)?.1;
    Some(rest.split_once('"')?.0.to_string())
}

fn validate_barcodes(barcodes: Vec<String>) -> Result<Vec<String>> {
    let mut seen = HashSet::new();
    for bc in &barcodes {
        if bc.len() != 10 {
            bail!("Found non-10bp barcode example: {bc}");
        }
        if !bc.bytes().all(|b| matches!(b, b'A' | b'C' | b'G' | b'T')) {
            bail!("Found barcode with non-ACGT bases example: {bc}");
        }
        if !seen.insert(bc.clone()) {
            bail!("Found duplicate barcode example: {bc}");
        }
    }
    Ok(barcodes)
}

fn write_whitelists(
    barcodes: &[String],
    middle_3p: &str,
    middle_5p: &str,
    out_3p: &Path,
    out_5p: &Path,
) -> Result<(usize, usize)> {
    let barcodes_5p: Vec<String> = barcodes.iter().map(|bc| reverse_complement(bc)).collect();
    if barcodes_5p.iter().collect::<HashSet<_>>().len() != barcodes_5p.len() {
        bail!("5p barcode reverse-complement list has duplicates.");
    }
    let mut writer_3p = File::create(out_3p)?;
    let mut count_3p = 0usize;
    for left in barcodes {
        for right in barcodes {
            writeln!(writer_3p, "{left}{middle_3p}{right}")?;
            count_3p += 1;
        }
    }

    let mut writer_5p = File::create(out_5p)?;
    let mut count_5p = 0usize;
    for left in &barcodes_5p {
        for right in &barcodes_5p {
            writeln!(writer_5p, "{left}{middle_5p}{right}")?;
            count_5p += 1;
        }
    }
    Ok((count_3p, count_5p))
}
